homeassistant/components/anomaly/binary_sensor.py

```diff
@@ -0,0 +1,326 @@
+        self.samples: deque = deque(maxlen=max_samples)
+
+    @property
+    def name(self) -> str:
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def is_on(self) -> str:
+        """Return true if sensor is on."""
+        return self._state
+
+    @property
+    def device_class(self) -> str:
+        """Return the sensor class of the sensor."""
+        return self._device_class
+
+    @property
+    def device_state_attributes(self) -> dict:
+        """Return the state attributes of the sensor."""
+        return {
+            ATTR_SAMPLE_COUNT: len(self.samples),
+            ATTR_TRAILING_SAMPLE_COUNT: len(self.trailing_samples),
+            ATTR_SAMPLE_AVG: self._sample_avg,
+            ATTR_TRAILING_AVG: self._trailing_avg,
+            ATTR_CHANGE_AMOUNT: self._change_amount,
+            ATTR_CHANGE_PERCENT: self._change_percent,
+        }
+
+    @property
+    def should_poll(self) -> bool:
+        """No polling needed."""
+        return False
+
+    async def async_added_to_hass(self) -> None:
+        """Complete device setup after being added to hass."""
+
+        @callback
+        def anomaly_sensor_state_listener(event) -> None:
+            """Handle state changes on the observed device."""
+            new_state: State = event.data.get("new_state")
+            if new_state is None:
+                return
+            try:
+                if self._attribute:
+                    state: str = new_state.attributes.get(self._attribute)
+                else:
+                    state: str = new_state.state
+                if state not in (STATE_UNKNOWN, STATE_UNAVAILABLE):
+                    sample: tuple = (new_state.last_updated.timestamp(), float(state))
+                    self.samples.append(sample)
+                    self.trailing_samples.append(sample)
+                    self.async_schedule_update_ha_state(True)
+            except (ValueError, TypeError) as ex:
+                _LOGGER.error(ex)
+            return
+
+        self.async_on_remove(
+            async_track_state_change_event(
+                self.hass, [self._entity_id], anomaly_sensor_state_listener
+            )
+        )
+        return
+
+    async def async_update(self) -> None:
+        """Get the latest data and update the states."""
+        # Remove outdated samples
+        if self._trailing_sample_duration > 0:
+            await self.hass.async_add_executor_job(self._trim_trailing_samples)
+        if self._sample_duration > 0:
+            await self.hass.async_add_executor_job(self._trim_samples)
+
+        # Calculate diffs between trailing and sample averages
+        if len(self.samples) == 0:
+            await self.hass.async_add_executor_job(self._empty_sample_diff)
+        else:
+            await self.hass.async_add_executor_job(self._calculate_diff)
+
+        # Update state
+        await self.hass.async_add_executor_job(self._set_state)
```

Comment:
Why are all these operations run in the thread pool? They don't look slow. Are they?

Reply:
I made some quick timing and calculations and it looks like we can sum a list with about 5 million items before it takes longer than 0.1 s.

A worst case scenario where the target sensor changes every second it would mean about 60 days worth of data. Is this a good estimate?

Reply:
Wasn't thinking. The async_update is already its own thread. I will remove those.